### PR TITLE
Correct access rights check for `RESTORE`

### LINKED ISSUE #71853: RESTORE requires more permissions than necessary
Backups can contain users, roles and grants. When restoring such a backup, clickhouse checks if the current user has the required permissions to create the users,roles and grants in the backup. However this check is too strict.

Assume the user `default` has the following grants
```
GRANT SELECT ON *.* TO default WITH GRANT OPTION
REVOKE SELECT ON system.zookeeper FROM default
```
(so the user can read anything except system.zookeeper).

This user can create another role `foobar`:
```
CREATE ROLE foobar;
GRANT CURRENT GRANTS to foobar;
```

If we now take a backup, then the `access-01.txt` would contain the following code grants for user `foobar`:
```
ATTACH ROLE foobar;
ATTACH GRANT SELECT ON *.* TO foobar;
ATTACH REVOKE SELECT ON system.zookeeper FROM foobar;
```

**If the `default` user tries to restore the backup, it will fail with a `Not enough privileges.` error, because the `default` user is not allowed to run `ATTACH GRANT SELECT ON *.* TO foobar WITH GRANT OPTION;`**

Can we change the way grants are restored, so that this example would work (and so that the  default user can successfully restore this backup with the `foobar` role)? In theory it should be possible. The `default` user was the user who created the `foobar` role in the first place. So it should also be allowed to restore it from backup.

We currently do the permission check for backups in `RestorerFromBackup::checkAccessForObjectsFoundInBackup` using `AccessRestorerFromBackup::getRequiredAccess`. This [function](https://github.com/ClickHouse/ClickHouse/blob/master/src/Access/AccessBackup.cpp) walks through all lines in `access-01.txt`. That's why currently in the above case, the restoring user needs `GRANT SELECT ON *.* WITH GRANT OPTION` (which it doesn't have, because it's missing the `SELECT ON system.zookeeper`).

We could implement a more accurate access checking, because the current implementation just skips partial revokes ([code](https://github.com/ClickHouse/ClickHouse/blob/9caac43b2aa5e7c5474a87b7596dea95f5a2569a/src/Access/AccessBackup.cpp#L361-L362)).

### PR DESCRIPTION
**Pull request by pufit**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
`RESTORE` operations for access entities required more permission than necessary because of unhandled partial revokes. This PR fixes the issue.
Closes https://github.com/ClickHouse/ClickHouse/issues/71853

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 24ad5462b2fc5ffc4027742ff992e149c9a32ca3 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/71958/24ad5462b2fc5ffc4027742ff992e149c9a32ca3/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/64712/d4599a68fa0ab5f85311d1af6cb27b9f866a82bf/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: tests/integration/test_backup_restore_new/test.py [hunk 1]
```diff
@@ -1845,6 +1846,30 @@ def drop():
...
+    instance.query("DROP USER u1")
+
+    instance.query("CREATE USER u2")
+    instance.query("GRANT SELECT ON *.* TO u2 WITH GRANT OPTION")
+    instance.query("GRANT CREATE USER ON *.* TO u2")
+    instance.query("REVOKE SELECT ON system.zookeeper* FROM u2")
+    instance.query("REVOKE SELECT ON foo.* FROM u2")
+
+    instance.query(f"RESTORE ALL FROM {backup_name}", user="u2")
+    instance.query("DROP USER u1")
```
Comment (vitlibar) [inline tests/integration/test_backup_restore_new/test.py:1866]: What is the point of dropping user u1 here?

File: tests/integration/test_backup_restore_new/test.py [hunk 2]
```diff
@@ -1845,6 +1846,30 @@ def drop():
...
+    instance.query("REVOKE SELECT ON foo* FROM u2")
+
+    with pytest.raises(QueryRuntimeException):
+        instance.query(f"RESTORE ALL FROM {backup_name}", user="u2")
```
Comment (vitlibar) [inline tests/integration/test_backup_restore_new/test.py:1870]: It's better to write here in the test what exception you expect.

File: tests/integration/test_backup_restore_new/test.py [hunk 3]
```diff
@@ -1845,6 +1846,30 @@ def drop():
...
+
+    instance.query("CREATE USER u2")
+    instance.query("GRANT SELECT ON *.* TO u2 WITH GRANT OPTION")
+    instance.query("GRANT CREATE USER ON *.* TO u2")
+    instance.query("REVOKE SELECT ON system.zookeeper* FROM u2")
+    instance.query("REVOKE SELECT ON foo.* FROM u2")
+
+    instance.query(f"RESTORE ALL FROM {backup_name}", user="u2")
+    instance.query("DROP USER u1")
+    instance.query("REVOKE SELECT ON foo* FROM u2")
```
Comment (vitlibar) [inline tests/integration/test_backup_restore_new/test.py:1867]: `foo.*` was already revoked from `u2` before `instance.query(f"RESTORE ALL FROM {backup_name}", user="u2")`, what's the point of repeating the same command here?

File: src/Backups/RestorerFromBackup.cpp [hunk 4]
```diff
@@ -704,9 +705,14 @@ void RestorerFromBackup::checkAccessForObjectsFoundInBackup() const
...
 
-    /// We convert to AccessRights and back to check access rights in a predictable way
-    /// (some elements could be duplicated or not sorted).
-    required_access = AccessRights{required_access}.getElements();
+    /// We convert to AccessRights to check if we have the rights contained in our current access.
+    /// This handles the situation when restoring partial revoked grants:
+    /// GRANT SELECT ON *.*
+    /// REVOKE SELECT FROM systems.*
+    const auto & required_access_rights = AccessRights{required_access};
+    const auto & access_rights = context->getAccess()->getAccessRights();
```
Comment (vitlibar) [inline src/Backups/RestorerFromBackup.cpp:713]: I'd rename `access_rights` to `current_user_access_rights` to make the difference with `required_access_rights` clearer.

File: src/Backups/RestorerFromBackup.cpp [hunk 5]
```diff
@@ -704,9 +705,14 @@ void RestorerFromBackup::checkAccessForObjectsFoundInBackup() const
...
+    if (access_rights->contains(required_access_rights))
```
Comment (vitlibar) [inline src/Backups/RestorerFromBackup.cpp:714]: Shouldn't it be
```
if (access_rights->containsWithGrantOption(required_access_rights))
```
?

File: src/Backups/RestorerFromBackup.cpp [hunk 6]
```diff
@@ -704,9 +705,14 @@ void RestorerFromBackup::checkAccessForObjectsFoundInBackup() const
...
+        return;
 
     context->checkAccess(required_access);
```
Comment (vitlibar) [inline src/Backups/RestorerFromBackup.cpp:717]: Do we need this `context->checkAccess(required_access)` here only to generate an error message?

File: src/Access/AccessRights.cpp [hunk 7]
```diff
@@ -1457,18 +1463,18 @@ bool AccessRights::isGrantedImpl(const AccessFlags & flags, const Args &... args
...
             return true;
-        return root_node->contains(*other.root);
+        return root_node->contains(*other_root_node);
     };
     if constexpr (grant_option)
-        return helper(root_with_grant_option);
+        return helper(root_with_grant_option, other.root);
     else
-        return helper(root);
+        return helper(root, other.root) && helper(root_with_grant_option, other.root_with_grant_option);
```
Comment (vitlibar) [inline src/Access/AccessRights.cpp:1477]: That look wrong, shouldn't it be
```
    if constexpr (grant_option)
        return helper(root_with_grant_option, other.root_with_grant_option);
    else
        return helper(root, other.root);
```
?

File: src/Backups/RestorerFromBackup.cpp (hunk 5 above)
Comment (vitlibar) [reply src/Backups/RestorerFromBackup.cpp:714]: Ok, I see it now.

File: src/Access/AccessRights.cpp (hunk 7 above)
Comment (vitlibar) [reply src/Access/AccessRights.cpp:1477]: Ok, I see it now, but let's write some comments about this behavior before the declaration of `AccessRights::contains()` and `AccessRights::containsWithGrantOption()`.

File: src/Backups/RestorerFromBackup.cpp (hunk 6 above)
Comment (vitlibar) [inline src/Backups/RestorerFromBackup.cpp:717]: Maybe we could make `contains` a part of `checkAccess`, WDYT?

Comment (pufit) [reply src/Backups/RestorerFromBackup.cpp:714]: Yes

Comment (pufit) [reply src/Backups/RestorerFromBackup.cpp:714]: I tried implementing this, but the code change has grown beyond the current PR. I will continue the refactoring in the separate PR, but this required a complete rewriting of `checkAccessImplHelper` and some other functions from checking access per flag to check access per access_rights set

File: tests/integration/test_backup_restore_new/test.py (hunk 1 above)
Comment (pufit) [reply tests/integration/test_backup_restore_new/test.py:1899]: To run the restore sequence, two lines below

File: tests/integration/test_backup_restore_new/test.py (hunk 3 above)
Comment (pufit) [reply tests/integration/test_backup_restore_new/test.py]: It's a different permission. I made it more apparent now

**Review by vitlibar — APPROVED** [review]

**Pull request merged**